### DISCUSSION TITLE
Inline script not loading

### DISCUSSION THREAD

**Discussion by mohdlatif** [Baroshem/nuxt-security#302] [Q&A] [answered]
After enabling the addon, my custom script is not loading,

my script inside a component

```
<script setup>
const nonce = useNonce()
useHead({
  script: [
    {
      src: "/js/sd.js",
      type: "text/javascript",
      async: true,
      body: false,
      nonce: nonce,
      onload: () => {
        let selectDropdowns = document.querySelectorAll("select")
        selectDropdowns.forEach((dropdown) => {
          new CustomSelect({
            elem: dropdown,
            bubbles: true,
          })
        })
      },
      onloadstart: () => {},
    },
    // {
    //   mode: "server",
    //   // Set the mode to 'server' if you want to avoid re-loading the script on the client
    //   // Set the mode to 'client' if you want client-side only loading
    //   // see: https://github.com/unjs/unhead/issues/136
    // },
  ],
})
```

my nuxt config script 

```
...
headers: {
      contentSecurityPolicy: {
        "img-src": ["'self'", "data:", "https:"],
        "script-src": [
          // "'self'", // Fallback value, will be ignored by most modern browsers (level 3)
          // "https:", // Fallback value, will be ignored by most modern browsers (level 3)
          "'unsafe-inline'", // Fallback value, will be ignored by almost any browser (level 2)
          // "'strict-dynamic'", // Strict CSP via 'strict-dynamic', supported by most modern browsers (level 3)
          // "'nonce-{{nonce}}'", // Enables CSP nonce support for scripts in SSR mode, supported by almost any browser (level 2)

          "https://static.cdn.prismic.io", // Prismic script source
          "https://ixvxiuxv.prismic.io", // Your self-hosted script source
        ]
....
```

when i view the page, the sd.js is loading however the scripts that is within onload not running 

I get this error

`Refused to execute inline event handler because it violates the following Content Security Policy directive: "script-src-attr 'none'". Either the 'unsafe-inline' keyword, a hash ('sha256-...'), or a nonce ('nonce-...') is required to enable inline execution. Note that hashes do not apply to event handlers, style attributes and javascript: navigations unless the 'unsafe-hashes' keyword is present.`

**Comment by Baroshem** [comment]
Hey,

Check out this documentation page :) 

https://nuxt-security.vercel.app/documentation/advanced/strict-csp#whitelisting-external-resources

**Comment by Baroshem** [comment]
And maybe our CSP expert @vejja could help here :)

**Comment by vejja** [comment]
Following the error recommendation in the console will do it: `"script-src-attr": ["'unsafe-inline'"]`

Additional Notes
#### 1- Unsafe-hashes
If you want to go one step further, you can also use the `'unsafe-hashes'` policy, which requires you to manually hash the content of the `onload` inline handler. A quick look at the [unjs/unhead github code](https://github.com/unjs/unhead/blob/9af7e289812e49d170b583dbb67402c803e310f6/packages/unhead/src/plugins/eventHandlers.ts#L12) shows that the generated inline event handler is `this.dataset.onload = true`. I hashed this string manually and it gave me the following policy: `"script-src-attr": ["'unsafe-hashes'", "'sha256-jp2rwKRAEWWbK5cz0grQYZbTZyihHbt00dy2fY8AuWY='"]`. But beware because I have no guarantee that `unjs/unhead` will keep this syntax constant over release cycles so this looks more like a temporary fix to me than a permanent solution.
@Baroshem: Maybe you want to ask the Nuxt core team about their views on this. Depending on their feedback we might be able to automate the generation of unsafe inline event handlers hashes for our users ?

#### 2- CSP disabled
Please note that **Strict CSP is not enabled on your site**, because
- you have `"script-src" : ["'unsafe-inline'"]`
- you don't have the `nonce-{{nonce}}` placeholders
- setting the nonce via useNonce() in useHead is useless.

In short, you are allowing unsafe inline scripts and unsafe inline event handlers, so maybe you don't need CSP at all ?

**Reply by mohdlatif** [reply]
Thank you for your in-depth reply. I have tried the `unsafe inline` option and indeed solves the issues for me by adding it in Nuxt config file

```
 "script-src-attr": [
          "'self'", // Fallback value, will be ignored by most modern browsers (level 3)
          // "https:", // Fallback value, will be ignored by most modern browsers (level 3)
          "'unsafe-inline'", // Fallback value, will be ignored by almost any browser (level 2)
          // "'strict-dynamic'", // Strict CSP via 'strict-dynamic', supported by most modern browsers (level 3)
          // "'nonce-{{nonce}}'",
        ]
```

Honestly, I just started learning Nuxt early this week, I do not have much of an expert around generating hashes or nonce.

How would you go if you want to nonce this script while having Strict CSP since using `useNonce()` is useless

a clear example would really help me.

```
useHead({
  script: [
    {
      src: "/js/sd.js",
      type: "text/javascript",
      body: false,
      nonce,
      onload: () => {
        let selectDropdowns = document.querySelectorAll("select")
        selectDropdowns.forEach((dropdown) => {
          new CustomSelect({
            elem: dropdown,
            bubbles: true,
          })
        })
      },
    },
  ],
})
```


Thank you

**Reply by vejja** [reply]
Hi @mohdlatif 
I would do the following:

- nuxt.config.ts:
```
...
headers: {
      contentSecurityPolicy: {
        // do not modify other defaults, only script-src-attr
        "script-src-attr": [
          "'unsafe-hashes'",
          "'sha256-jp2rwKRAEWWbK5cz0grQYZbTZyihHbt00dy2fY8AuWY='"
        ]
      }
....
```

- Component code
```
<script setup>
// do not use useNonce, do not set nonce in useHead. This will be done automatically by Nuxt-Security
useHead({
  script: [
    {
      src: "/js/sd.js",
      type: "text/javascript",
      body: false,
      onload: () => {
        let selectDropdowns = document.querySelectorAll("select")
        selectDropdowns.forEach((dropdown) => {
          new CustomSelect({
            elem: dropdown,
            bubbles: true,
          })
        })
      },
    },
  ],
})
...
</setup>
```


I can see that you are using Prismic but I must say I never used them so I do not know whether they are CSP-compatible.

Content Security Policy is a very strict set of rules that prevent third-party scripts from running on your website. Nuxt Security implements Strict CSP Level 3, which is the highest CSP protection standard, so if you are using Nuxt Security with Prismic, there is a chance that Prismic scripts are considered 3rd party and won't load. 

In that case, there are workarounds, and it depends on your goals:
1- Do you need CSP ? If you don't, you can disable CSP protection in Nuxt Security, either completely or partially
2- If you need CSP, can you try the setup that I am recommending above, and tell me what the console says? I will be able to further guide you on setting up CSP properly
3- Can you share a little bit more on how Prismic deploys assets? If you have a repro (Github, Stackblitz) and/or a link to Prismic sources or deployment, this would be super helpful for me to understand how we can help other Prismic users.

Thanks

**Reply by Baroshem** [reply]
Hey @mohdlatif 

Have you tried the recommendations from @vejja ?

**Reply by mohdlatif** [reply, marked as answer]
Yeb, it worked, this is my config.



```
  security: {
    nonce: true,
    ssg: {
      hashScripts: true, // In the SSG case, inline scripts generated by the server will be allowed by hash
      hashStyles: false, // Disables CSP hash support for styles in SSG mode (recommended)
    },
    //https://nuxt-security.vercel.app/documentation/advanced/strict-csp
    headers: {
      contentSecurityPolicy: {
        "img-src": [
          "'self'",
          "data:",
          "https://res.cloudinary.com",
      
        ],
        "script-src": [
          "'self'",
          "'nonce-{{nonce}}'",
          "https://static.cdn.prismic.io",
          "https://domain.prismic.io",
        ],
        "script-src-attr": [
          "'unsafe-hashes'",
          "'sha256-jp2rwKRAEWWbK5cz0grQYZbTZyihHbt00dy2fY8AuWY='",
        ],
        "style-src": ["'self'", "'unsafe-inline'"],
        "frame-src": ["'self'", "https://domain.prismic.io"],

        "worker-src": ["'self'", "blob:"],
        "connect-src": [
          "'self'",
          "https://api.iconify.design",
          "https://api.unisvg.com",
          "https://domain.cdn.prismic.io",
        ],
        "upgrade-insecure-requests": true,
      },
      crossOriginOpenerPolicy: false,
      crossOriginEmbedderPolicy: false,
      xXSSProtection: "1; mode=block",
    },
  },
```

Thank you @Baroshem  @vejja

**Reply by Baroshem** [reply]
Awesome,

Kudos especially to @vejja 🚀

I am mainly a maintainer recently ;)

**Reply by Baroshem** [reply]
@vejja 

Do you think it would be useful to add a note about this question/issue in our docs? :)

**Reply by vejja** [reply]
Yes definitely

Although I’d like to ask @mohdlatif :
- Did you try ‘strict-dynamic’ on script-src ?
- Would you be willing to tell what kind of scripts you needed to load from static.cdn.prismic.io and domain.prismic.io ?

I am wondering whether ‘strict-dynamic’ was causing issues with these scripts

**Reply by mohdlatif** [reply]
Sure @vejja, 
### For [`strict-dynamic`](https://content-security-policy.com/strict-dynamic/) I have not really tried.  I just tried to modify my script-src from 

```
"script-src": [
          "'self'",
          "'nonce-{{nonce}}'",
          "https://static.cdn.prismic.io",
          "https://domain.prismic.io",
        ],
```

to 
```
 "script-src": [
          "'self'",
          "'strict-dynamic'", // Strict CSP via 'strict-dynamic', supported by most modern browsers (level 3)
          "'nonce-{{nonce}}'",
        ],
```

And Prismic loads fine, but there is a catch. [Not all browsers support](https://caniuse.com/?search=strict-dynamic) `CSP Level 3`

![image](https://github.com/Baroshem/nuxt-security/assets/56746772/b2facdaa-daf8-43fa-892b-d8ea6a6f7083)

So, there is a risk, that Prismic or external scripts not loads on old browser.

---
### When commenting out 

```
// crossOriginOpenerPolicy: false,
// crossOriginEmbedderPolicy: false,
```
I got this error 
![image](https://github.com/Baroshem/nuxt-security/assets/56746772/f71a8ccd-30c4-4f88-98e2-3ae238ebbbd0)

---


### For the 2nd question: basically you need to exclude 2 scripts, otherwise, Prismic content will not show on the page. All content fetched from prisimic.io will be shown as `undefined `on the page.

both scripts to exclude are mandatory; `static.cdn.prismic.io` is fixed
domain.prismic.io needs to be updated with your Prismic custom domain. `custom-domain`.prismic.io. It is created once you started a project.

**Reply by vejja** [reply]
Hi @mohdlatif and thanks for the reply

Would you be able to modify your config to :
```
"script-src": [
          "'self'",
          "'nonce-{{nonce}}'",
          "https://static.cdn.prismic.io",
          "https://domain.prismic.io",
          "'strict-dynamic'"
        ],
```
If a browser does not support CSP Level 3, it will fallback to your configuration. So this will safely work on all browsers, even the older ones.

And let me know if everything works correctly ? It would be super helpful !

**Reply by Baroshem** [reply]
I have na idea

@mohdlatif maybe you would be interested in contributing to our documentation (FAQ) section where you could add a record with using Nuxt Security and Prismic? Explaining your issue and how you have solved with a link to this discussion? I think it could be really useful for the community :)

**Reply by mohdlatif** [reply]
Version: "nuxt-security": "^1.0.0-rc.5",
I used the suggested code you provide and I think I spotted an issue w/ pages that set to true prerender.

The script in my `Nuxt Config`  has the following code, to help prerender some pages in advance, I am basically use it, so Vercel can cache my page and get HIT response 

```
  routeRules: {
    "/about": { static: true },
    "/api/**": { static: false, cache: false },
    "/test/pre": { prerender: true },
}
```

I have to mention that I am using i18n; English set as the alternate language

when navigating to domain.com/test/pre

I get this error

![image](https://github.com/Baroshem/nuxt-security/assets/56746772/c85e5ee7-288f-433b-976a-ff34de73588a)
![image](https://github.com/Baroshem/nuxt-security/assets/56746772/d934cc4f-4d0b-4944-994a-70b2718d772b)


however, navigating to `domain.com/en-us/test/pre`

No error presents.


Why do you think this issue happens ?

**Reply by mohdlatif** [reply]
Also I noticed that visiting the prerender page, a new meta tag is added.

`<meta http-equiv="Content-Security-Policy" content="base-uri 'none'; font-src 'self' https: data:; form-action 'self'; frame-ancestors 'self'; img-src 'self' data: https://res.cloudinary.com https://naive-ui.oss-cn-beijing.aliyuncs.com; object-src 'none'; script-src-attr 'unsafe-hashes' 'sha256-jp2rwKRAEWWbK5cz0grQYZbTZyihHbt00dy2fY8AuWY='; style-src 'self' 'unsafe-inline'; script-src 'self' https://static.cdn.prismic.io https://domain.prismic.io 'sha256-VdzDPWVNoiO30r0Jip1QLqU+ZQB/WbOHeEzwrqA+M7k=' 'sha256-9Oscc5BBwxJCa+clgTsSxDh+bAyTT8110OK0+j5i2XA=' 'sha384-4eKcb5xdb0rEsgZkKyEjRwcbXzNnsGqnYO0xFyGTwbRRoA6fMi+HuLyemzLRAx8q' 'sha384-md0YO9GpW8yfEgrarIweZNCbMLhzYTc7Cd40uzCqF74I/0CbjPphrdRySlsDfEcE' 'sha384-NJE29GQBcRMel8RFYMj7R1U2ZimlL/CGyP+6GKoSNu4VFvSXWsNqXmzOfrdpjO4/' 'sha256-co0sieW+l4L7Run6KX8zkkkJXxlTw0lofV2r5fW1c2o='; upgrade-insecure-requests; frame-src 'self' https://domain.prismic.io; worker-src 'self' blob:; connect-src 'self' https://api.iconify.design https://api.unisvg.com https://domain.cdn.prismic.io;">`


and this issue shows in console 

![image](https://github.com/Baroshem/nuxt-security/assets/56746772/f2f10f68-894d-40cf-8c61-653893a718eb)

**Reply by vejja** [reply]
Hi @mohdlatif and thanks so much for doing the tests here. We will be able to produce a detailed FAQ section for other Prismic users.

1- It is normal that CSP Level 3 blocks pre-rendered pages on `/test/pre`, because Prismic does not provide integrity hashes as far as I know.
The solution is to downgrade to CSP Level 2 for pre-rendered pages:
```
routeRules: {
  "/about": { static: true },
  "/api/**": { static: false, cache: false },
  "/test/pre": { 
    prerender: true,
    security: {
      headers: {
        contentSecurityPolicy: {
          'script-src': [
            "'self'",
            "https://static.cdn.prismic.io",
            "https://domain.prismic.io"
          ]
        }
      }
    }
  },
}
```

Unless you absolutely need to pre-render, I would advise to not use the `prerender: true` option if you can. CSP is much easier to deal with in SSR mode.

2- Your `/en-us/test/pre` page is not pre-rendered because it is not explicitly defined in `routeRules`.

The solution is to add: 
```
routeRules: {
  "/about": { static: true },
  "/api/**": { static: false, cache: false },
  "/test/pre": { ... },
  "/en-us/test/pre": { ... }
```

3- It is normal that the `http-equip` meta tag is added to your HTML, as this is the standard way of transmitting CSP policies for pre-rendered pages.

4- The `frame-ancestors` policies is not supported for pre-rendered page via meta. This is a known limitation of CSP, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors

5- The COOP/COEP issue that you mentioned in your earlier message is related to how Prismic applies CORS rules when fetching sources; this is unrelated to us but I will open an issue with Prismic because they can fix this easily.


Thank you again very much for your contribution. I think we can close this discussion, unless you have other questions ?

**Reply by mohdlatif** [reply]
Yes, lets close it. I want to say a big thanks for your help. I appreciate the time you take to respond to my questions.

**Comment by vejja** [comment]
Note: Prismic COOP/COEP Github Issue opened at : https://github.com/nuxt-modules/prismic/issues/204
